### PR TITLE
{monitor}`az monitor autoscale`: Migrate from codegen V2

### PR DESCRIPTION
**Pull request by AllyW**
**Related command**

az monitor autoscale create/delete/show/list
az monitor autoscale profile create/update/delete/show/list
az monitor autoscale rule create/delete/list/copy

**Description**

**Testing Guide**

**History Notes**

[Monitor] `az monitor autoscale`: Migrate from codegen V2

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
monitor

File: src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py [hunk 1]
```diff
@@ -3,16 +3,128 @@
...
+            self.ctx.vars.instance.properties.name = AAZUndefined
+        result = self.deserialize_output(self.ctx.vars.instance, client_flatten=True)
+        return result
+
+
+class AutoScaleShow(_AutoScaleShow):
+
+    def _output(self, *args, **kwargs):
+        from azure.cli.core.aaz import AAZUndefined
+        if has_value(self.ctx.vars.instance.properties.name):
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py:36]: Please add code comments here

File: src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py [hunk 2]
```diff
@@ -3,16 +3,128 @@
...
+        autoscale_name = parse_resource_id(resource)['name']
+    min_count = min_count or count
+    max_count = max_count or count
+    args = {}
+    default_profile = {
+        "name": DEFAULT_PROFILE_NAME,
+        "capacity": {
+            "default": str(count),
+            "minimum": str(min_count),
+            "maximum": str(max_count)
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py:83]: The value is string?

File: src/azure-cli/azure/cli/command_modules/monitor/commands.py [hunk 3]
```diff
@@ -164,24 +164,27 @@ def load_command_table(self, _):
...
     with self.command_group('monitor autoscale profile', autoscale_sdk, custom_command_type=autoscale_custom) as g:
-        g.custom_command('create', 'autoscale_profile_create')
-        g.custom_command('list', 'autoscale_profile_list')
-        g.custom_show_command('show', 'autoscale_profile_show')
-        g.custom_command('delete', 'autoscale_profile_delete')
-        g.custom_command('list-timezones', 'autoscale_profile_list_timezones')
+        g.custom_command('create', 'autoscale_profile_create_new')  # done
+        g.custom_command('list', 'autoscale_profile_list')  # done
+        g.custom_show_command('show', 'autoscale_profile_show')  # done
+        g.custom_command('delete', 'autoscale_profile_delete_new')  # done
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/commands.py:180]: You can use sub commands for profile.
image

File: src/azure-cli/azure/cli/command_modules/monitor/commands.py [hunk 4]
```diff
@@ -164,24 +164,27 @@ def load_command_table(self, _):
...
 
     with self.command_group('monitor autoscale rule', autoscale_sdk, custom_command_type=autoscale_custom) as g:
-        g.custom_command('create', 'autoscale_rule_create')
-        g.custom_command('list', 'autoscale_rule_list')
-        g.custom_command('delete', 'autoscale_rule_delete')
-        g.custom_command('copy', 'autoscale_rule_copy')
+        g.custom_command('create', 'autoscale_rule_create_new')  # done
+        g.custom_command('list', 'autoscale_rule_list')  # done
+        g.custom_command('delete', 'autoscale_rule_delete_new')  # done
+        g.custom_command('copy', 'autoscale_rule_copy_new')  # done
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/commands.py:187]: You can also create sub commands for the rules as well.

File: src/azure-cli/azure/cli/command_modules/monitor/_autoscale_util.py [hunk 5]
```diff
@@ -557,6 +557,87 @@
...
+                    profile_schedule[day][sched_name]['capacity'] = profile.capacity
+                    profile_schedule[day][sched_name]['rules'] = profile.rules
+                else:
+                    profile_schedule[day][sched_name] = {
+                        'start': start_time,
+                        'capacity': profile.capacity,
+                        'rules': profile.rules
+                    }
+
+    return default_profile, profile_schedule
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/_autoscale_util.py:638]: The default_profile here can be a Dict or a AAZObject. They are different.

File: src/azure-cli/azure/cli/command_modules/monitor/_autoscale_util.py [hunk 6]
```diff
@@ -662,3 +743,112 @@ def _profile_is_subset(profile, start, end):
             logger.warning("Proposed schedule '%s %s-%s' has a full or partial overlap with the following existing "
                            "schedules: %s. Unexpected behavior may occur.",
                            day, start, end, ', '.join(profile_conflicts))
+
+
+def build_autoscale_profile_dict(autoscale_settings):
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/_autoscale_util.py:748]: It is possible keep either `build_autoscale_profile_dict` or `build_autoscale_profile_from_instance`

File: src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py [hunk 7]
```diff
@@ -279,50 +486,104 @@ def autoscale_profile_list_timezones(cmd, client, offset=None, search_query=None
...
+        new_default, _ = build_autoscale_profile_from_instance(instance)
+        if not new_default:
+            instance.properties.profiles.append(default_profile)
+
+    def _output(self, *args, **kwargs):
+        from azure.cli.core.aaz import AAZUndefined
+        if has_value(self.ctx.vars.instance.properties.name):
+            self.ctx.vars.instance.properties.name = AAZUndefined
+        result = self.deserialize_output(self.ctx.vars.instance, client_flatten=True)
+        return result
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py:536]: Do we need output for delete command?

File: src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py [hunk 8]
```diff
@@ -279,50 +486,104 @@ def autoscale_profile_list_timezones(cmd, client, offset=None, search_query=None
...
+
+
+def autoscale_profile_delete(cmd, autoscale_name, resource_group_name, profile_name):
+    AutoScaleProfileDelete(cli_ctx=cmd.cli_ctx)(command_args={
+        "autoscale_name": autoscale_name,
+        "resource_group": resource_group_name,
+        "profile_name": profile_name,
+    })
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py:544]: You should return it. Specially when it's a LRO

File: src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py [hunk 9]
```diff
@@ -175,96 +224,254 @@ def _parse_action_removals(actions):
...
+            usage = valid_recurrence[delimiter]['usage']
+            try:
+                args.recurrence_obj = valid_recurrence[delimiter]['validator'](args)
+            except ValidationError as ex:
+                raise ValidationError('{} invalid usage: {}'.format(ex, usage))
+        except KeyError:
+            raise ValidationError('invalid usage: -r {{{}}} [ARG ARG ...]'.format(','.join(valid_recurrence)))
+
+
+class AutoScaleProfileCreate(_AutoScaleUpdate):
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py:393]: Is this class used?

File: src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py [hunk 10]
```diff
@@ -175,96 +224,254 @@ def _parse_action_removals(actions):
...
+    from azure.cli.command_modules.monitor._autoscale_util import AUTOSCALE_TIMEZONES
+    value = args.timezone.to_serialized_data()
+    zone = next((x['name'] for x in AUTOSCALE_TIMEZONES if x['name'].lower() == value.lower()), None)
+    if not zone:
+        raise ValidationError(
+            "Invalid time zone: '{}'. Run 'az monitor autoscale profile list-timezones' for values.".format(value))
+    args.timezone = zone
+
+
+def validate_recurrence(args):
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py:347]: There can also be removed as `AutoScaleProfileCreate` class is not used.

File: src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py [hunk 11]
```diff
@@ -341,63 +602,118 @@ def preprocess_for_spring_cloud_service():
...
+            "value": scale.value
+        }
+    }
+    profile["rules"].append(rule)
+    updated_rets = AutoScaleUpdate(cli_ctx=cmd.cli_ctx)(command_args=autoscale_settings)
+    updated_profile = _identify_profile_cg(updated_rets["profiles"], profile_name)
     # determine if there are unbalanced rules
-    scale_out_rule_count = len([x for x in profile.rules if x.scale_action.direction == ScaleDirection.increase])
-    scale_in_rule_count = len([x for x in profile.rules if x.scale_action.direction == ScaleDirection.decrease])
+    scale_out_rule_count = len([x for x in profile["rules"] if x["scale_action"]["direction"] == "Increase"])
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py:620]: Do you need to use updated_profile here?

File: src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py [hunk 12]
```diff
@@ -341,63 +602,118 @@ def preprocess_for_spring_cloud_service():
...
-    autoscale_settings = client.get(resource_group_name, autoscale_name)
-    src_profile = _identify_profile(autoscale_settings.profiles, source_profile)
-    dst_profile = _identify_profile(autoscale_settings.profiles, dest_profile)
-    if '*' in index:
-        dst_profile.rules = src_profile.rules
-    else:
-        for i in index:
-            dst_profile.rules.append(src_profile.rules[int(i)])
-    autoscale_settings = client.create_or_update(resource_group_name, autoscale_name, autoscale_settings)
+    AutoScaleRuleCopy(cli_ctx=cmd.cli_ctx)(command_args={
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py:710]: Do you need to return it?

File: src/azure-cli/azure/cli/command_modules/monitor/_autoscale_util.py (hunk 5 above)
Comment (AllyW) [reply src/azure-cli/azure/cli/command_modules/monitor/_autoscale_util.py:636]: default_profile's format is accepted for instance property update

File: src/azure-cli/azure/cli/command_modules/monitor/_autoscale_util.py (hunk 6 above)
Comment (AllyW) [reply src/azure-cli/azure/cli/command_modules/monitor/_autoscale_util.py:746]: redesign after update migration done.

File: src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py (hunk 7 above)
Comment (AllyW) [reply src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py:404]: no need as previous func, ignored in wrapper

File: src/azure-cli/azure/cli/command_modules/monitor/aaz/latest/monitor/autoscale/_update.py [hunk 13]
```diff
@@ -0,0 +1,1108 @@
...
+@register_command(
+    "monitor autoscale update",
+)
+class Update(AAZCommand):
+    """Update an autoscale setting.
+
+    For more information on autoscaling, visit: https://docs.microsoft.com/azure/monitoring-and-diagnostics/monitoring-understanding-autoscale-settings.
+
+    :example: Update autoscale settings to use a fixed 3 instances by default.
+        az monitor autoscale update -g {myrg} -n {autoscale-name} --count 3
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/monitor/aaz/latest/monitor/autoscale/_update.py:23]: Please don't use placeholder characters in examples. You should use fake value in examples.
```suggestion
        az monitor autoscale update -g myrg -n autoscale-name --count 3
```

**Review by kairu-ms — DISMISSED** [review]

File: src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py (hunk 8 above)
Comment (AllyW) [reply src/azure-cli/azure/cli/command_modules/monitor/operations/autoscale_settings.py:411]: no need as previous func

**Review by necusjz — APPROVED** [review]

**Pull request merged**